rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,1222 @@
+basic `version-directive` that could be used to display the *recommended* version for
+teleport instances.
+
+In theory, the ability to display recommended version, and/or generate notifications, is
+a less "core" functionality, and could be added in a later step with less overall
+development effort. Once the TUF controller exists, using its output for notifications
+would be easy. That being said, it may be more valuable to deliver a pretty good way of
+informing users that they aught to upgrade sooner, rather than waiting on a very robust
+way of automatically upgrading that happens to bring notifications along with it.
+
+Regardless of ordering, notifications in general depend on the following components
+that need to be built anyway:
+
+- The target+server matching part of the `version-directive` resource (installer matching
+comes later).
+
+- The draft phase of the [version directive flow](#version-directive-flow).
+
+- The basic top-level config API (get/put/del).
+
+- The basic version controller configuration API (get/put/del) (only required if
+we want to support a controller configuration other than `default`).
+
+- The `tctl version-control status` command (though not all fields will be available yet).
+
+Creating the notification-only system first will also necessitate an additional
+builtin `version-controller` that would not otherwise be needed. Luckily, it can
+be *very* simple (e.g. a github release page scraper), since it will be explicitly
+*not* usable for actual upgrades.
+
+
+### Script-Based Upgrades MVP
+
+With the core work done for inventory status/control, we can move on to a barebones
+MVP/pathfinder for installers, version directives, and the version reconciliation loop.
+We will implement a no-frills version of these components with the goal of supporting
+one specific use-case: manually creating a basic `version-directive` resource and
+having a user-provided script run on all servers that don't match the directive.
+
+This phase will be a bit of a pathfinder, with a focus on weeding out any issues with
+the proposed design of the core system. It will also provide an early preview for users
+that are interested in reducing per-instance upgrade work, but are still willing to get
+their hands dirty. Finally, this will mark the point after which manual upgrade of
+non-auth instances can (theoretically) end, as new versions that support new installers
+can be "bootstrapped" using older installers.
+
+The components that must be developed for this phase are as follows:
+
+- Per-instance installation attempt status info.
+
+- The version reconciliation loop (minus more advanced features like being able to
+trigger remote installers).
+
+- The version directive resource (mostly complete already if we did the notification-only
+system first), and the version directive flow.
+
+- The `local-script` installer, and basic installer configuration API (get/put/del).
```

Comment:
Will we provide apt-based and rpm-based update scripts in their respective packages?

Reply:
Doable, I wouldn't want to mess with our package builds unless we really saw demand for it, and if we did it might be better to just have teleport come with "builtins" for specific package managers, that way we could do bug fixes/improvements behind the scenes.  To begin with, examples in the docs are probably simples/best.